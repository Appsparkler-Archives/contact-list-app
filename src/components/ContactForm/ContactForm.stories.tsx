import type { Meta, StoryObj } from "@storybook/react";
import { ContactForm } from "./ContactForm";

const meta = {
  title: "Components/Contact Form",
  component: ContactForm,
  tags: ["components"],
} satisfies Meta<typeof ContactForm>;

export default meta;

type Story = StoryObj<typeof meta>;

export const contactForm: Story = {
  args: {
    contact: {
      gender: "male",
      mobile: "+91-719-8101-355",
      name: "John Smith",
      type: "business",
      address: "Fathimapur",
      company: "PS",
      email: "appsparkler@gmail.com",
    },
    // contacts,
  },
};
