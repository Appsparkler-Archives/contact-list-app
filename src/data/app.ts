import { IContactFormData } from "types";

export type TAppFC = React.FC<{
  contacts: IContactFormData[];
}>;
