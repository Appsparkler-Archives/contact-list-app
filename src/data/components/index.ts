export * from "./standard-select-field";
