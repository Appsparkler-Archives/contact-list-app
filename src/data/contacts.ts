import { uniqueId } from "lodash/fp";
import { IContactFormData } from "types";

export const contactsData: Record<"variant1", IContactFormData[]> = {
  variant1: [
    {
      id: "7cb1248c-983a-4207-a59f-40bfa1fae564",
      name: "Ali Connors",
      gender: "male",
      mobile: "+9282822223",
      email: "ali.connors@whatsapp.com",
      type: "business",
      imgUrl:
        "https://plus.unsplash.com/premium_photo-1671641797679-3b680a7d2109?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MXx8cHJvZmlsZSUyMHBpY3xlbnwwfHwwfHw%3D&auto=format&fit=crop&w=900&q=60",
      company: "SurveyHeart",
      address: "UK",
    },
    {
      id: "b4fc7a3e-3522-4292-b05d-5edbdffcf3ac",
      name: "Jenny Allen",
      gender: "female",
      mobile: "+39292323",
      type: "personal",
      imgUrl:
        "https://images.unsplash.com/photo-1625621435205-2d09ed661a71?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=987&q=80",
      company: "Societe Generale",
    },
    {
      id: "e90ae941-f295-4adc-bf8a-b1d7127a60e6",
      gender: "male",
      mobile: "+9211111214343",
      type: "business",
      imgUrl:
        "https://images.unsplash.com/photo-1500648767791-00dcc994a43e?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8Mnx8cHJvZmlsZSUyMHBpY3xlbnwwfHwwfHw%3D&auto=format&fit=crop&w=900&q=60",
      name: "Steven Covey",
    },
  ],
};

export const getDefaultCreateContact = (): IContactFormData => ({
  id: uniqueId("contact"),
  gender: "female",
  mobile: "",
  name: "",
  type: "personal",
  address: "",
  company: "",
  email: "",
});
