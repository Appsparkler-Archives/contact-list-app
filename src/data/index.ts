export * from "./contacts";
export * from "./components";
