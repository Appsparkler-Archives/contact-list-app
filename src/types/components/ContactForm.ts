import { IContactFormData } from "types";

export type TContactFormFC = React.FC<{
  contact: IContactFormData;
  submitBtnTitle: string;
  formTitle: string;

  // event handlers
  onClickSubmit: (contactData: IContactFormData) => void;
  onClickCancel: () => void;
}>;
