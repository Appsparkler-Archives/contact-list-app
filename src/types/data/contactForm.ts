export interface IContactForm {
  name: string;
  mobile: string;
  address: string;
  company: string;
  email: string;
  gender: "male" | "female";
  type: "personal" | "business";
}
