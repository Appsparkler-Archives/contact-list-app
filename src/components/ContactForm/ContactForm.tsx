import { Box, Button, TextField, Typography } from "@mui/material";
import { TContactFormFC } from "types/components/ContactForm";
import { noop } from "lodash/fp";
import { StandardSelectField } from "components/StandardSelectField/StandardSelectField";
import { useCallback, useState } from "react";
import { IContactFormData, IStandardSelectFieldChangeHandler } from "types";

export const ContactForm: TContactFormFC = ({
  contact,
  submitBtnTitle,
  formTitle,
  onClickSubmit = noop,
  onClickCancel = noop,
}) => {
  const [contactData, setContactData] = useState<IContactFormData>(contact);

  const handleSubmit = useCallback<
    React.FormEventHandler<HTMLFormElement>
  >(() => {}, []);

  const handleChangeTextField = useCallback<
    React.ChangeEventHandler<HTMLInputElement>
  >(({ target: { name, value } }) => {
    setContactData((oldContactData) => ({
      ...oldContactData,
      [name]: value,
    }));
  }, []);

  const handleChangeSelectField =
    useCallback<IStandardSelectFieldChangeHandler>((name, value) => {
      setContactData((oldContactData) => ({
        ...oldContactData,
        [name]: value,
      }));
    }, []);

  return (
    <form onSubmit={handleSubmit}>
      <Box gap={2} flexDirection={"column"} display={"flex"}>
        <Typography variant="h5">{formTitle}</Typography>
        <TextField
          fullWidth
          label="Name"
          variant="standard"
          name="name"
          value={contactData.name}
          onChange={handleChangeTextField}
          required
        />
        <TextField
          fullWidth
          label="Mobile"
          variant="standard"
          name="mobile"
          value={contactData.mobile}
          onChange={handleChangeTextField}
          required
        />
        <TextField
          fullWidth
          label="Email"
          variant="standard"
          value={contactData.email}
          onChange={handleChangeTextField}
          name="email"
        />
        <StandardSelectField
          options={[
            { name: "Female", value: "female" },
            { name: "Male", value: "male" },
          ]}
          id="contact-field--gender"
          label="Gender"
          onChange={handleChangeSelectField}
          selectedOption={contactData.gender}
          name="gender"
        />
        <StandardSelectField
          options={[
            { name: "Personal", value: "personal" },
            { name: "Business", value: "business" },
          ]}
          id="contact-field--type"
          label="Type"
          onChange={handleChangeSelectField}
          selectedOption={contactData.type}
          name="type"
        />
        <TextField
          fullWidth
          label="Company"
          variant="standard"
          value={contactData.company}
          name="company"
          onChange={handleChangeTextField}
        />
        <Box
          display={"flex"}
          flexDirection={"row"}
          justifyContent={"space-around"}
          mt={3}
        >
          <Button size="large" variant="outlined">
            CANCEL
          </Button>
          <Button variant="contained" size="large" type="submit">
            {submitBtnTitle}
          </Button>
        </Box>
      </Box>
    </form>
  );
};
