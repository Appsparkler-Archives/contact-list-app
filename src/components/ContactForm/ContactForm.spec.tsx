import renderer from "react-test-renderer";
import { IContactFormData } from "types/data/contactForm";
import { noop } from "lodash/fp";
import { ContactForm } from "./ContactForm";

describe("ContactForm component", () => {
  test("match snapshot", async () => {
    const contact: IContactFormData = {
      name: "",
      gender: "male",
      mobile: "",
      type: "business",
    };
    const component = renderer.create(
      <ContactForm
        contact={contact}
        submitBtnTitle={"Edit"}
        formTitle={"Edit Contact"}
        onSubmit={noop}
        onCancel={noop}
      />
    );
    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
