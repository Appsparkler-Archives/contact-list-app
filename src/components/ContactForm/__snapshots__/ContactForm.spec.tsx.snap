// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm component match snapshot 1`] = `
<form>
  <div
    className="MuiBox-root css-y58rij"
  >
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r0:"
        id=":r0:-label"
      >
        Name
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          disabled={false}
          id=":r0:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        htmlFor=":r1:"
        id=":r1:-label"
      >
        Mobile
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          disabled={false}
          id=":r1:"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
        id="contact-field--gender"
      >
        Gender
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl  css-m5hdmq-MuiInputBase-root-MuiInput-root-MuiSelect-root"
        onClick={[Function]}
        required={true}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="contact-field--gender"
          className="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Male
        </div>
        <input
          aria-hidden={true}
          aria-invalid={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={true}
          tabIndex={-1}
          value="male"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
        id="contact-field--type"
      >
        Type
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl  css-m5hdmq-MuiInputBase-root-MuiInput-root-MuiSelect-root"
        onClick={[Function]}
        required={true}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="contact-field--type"
          className="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Personal
        </div>
        <input
          aria-hidden={true}
          aria-invalid={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={true}
          tabIndex={-1}
          value="personal"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</form>
`;
